### PR TITLE
Python 3 compatibility changes

### PR DESCRIPTION
**Pull request by chadheyne**
- Change import statements to from .x import y
- Change dict.iteritems() to dict.items()
- Use integer division in diagnostics 
- Force values to list in cases of dict.values() or map(f, list) or
  filter(f, list)
- Change xrange to range and use list(range) when code expects indices.
- All tests pass on Windows/Linux Python 3.3.2
- Additional minor pep8 changes throughout. 
- I don't have Python 2.x installed so can't test changes there. The only things that I think might be an issue are the changes to // instead of / and in Python 3 os.path.walk was moved to just os.walk.

(I've never done a pull request before and am new overall to contributing. Let me know if I need to change something or include more details)

### CODE REVIEW

**Comment by chadheyne** [comment]
Looks like the build fails because of the Python 3 change to metaclass definition. I'm not sure if **metaclass** still works in Python 3 or if this will have to include six.with_metaclass function

**Comment by twiecki** [comment]
@chadheyne This looks great!
The `glm/families.py` and `glm/links.py` metaclasses? I think we can just get rid of them as they only turn the class into an ABC.

**Comment by twiecki** [comment]
@chadheyne Also, can you enable python 3.3 in travis.yml?

**Comment by chadheyne** [comment]
I changed the travis.yml file to Python 3.3. The two metaclasses were families.Family and links.LinkFunction I believe.

**Comment by twiecki** [comment]
Yeah those you can just remove.

**Comment by chadheyne** [comment]
Not sure why travis is failing on TestGLM. I completely removed pymc from my system and reinstalled it and the tests still pass.

**Comment by twiecki** [comment]
Yeah, I disabled the tests on a separate branch. Also can't reproduce it locally. If the others pass this should be good to merge after code review.

Did you use 2to3 for this?

**Comment by chadheyne** [comment]
For some portions of it. 2to3 is quite pushy though and tries to change all instances of dict.values() to list(dict.values()) even if it's being used in an iterator context, etc. Basically I was just trying to read through a tutorial using pymc and just fixed things as I went along then after realizing how much I'd changed I figured I might as well fix up the tests to help others out when using Py3.

**Comment by chadheyne** [comment]
Do you need me to change/run anything else or will the changes on your other branch cover it?

**Comment by twiecki** [comment]
My changes should cover it.

On Mon, Nov 11, 2013 at 10:10 PM, chadheyne notifications@github.comwrote:

> Do you need me to change/run anything else or will the changes on your
> other branch cover it?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/pymc-devs/pymc/pull/385#issuecomment-28264519
> .

## 

Thomas Wiecki
PhD candidate, Brown University
Quantitative Researcher, Quantopian Inc, Boston

**Comment by twiecki** [comment]
Can you rebase off of master? This should make master travis pass again and we can make sure your PR doesn't break existing things.

**Comment by chadheyne** [comment]
I've never done rebasing before so I may have completely screwed stuff up. I had to make a couple changes in test_distributions after the rebase in order to pass tests locally with Python 3. 

(I am slightly OCD about whitespace so I tried to keep this file consistent with the others, you can ignore this if you want, though)

**Comment by chadheyne** [comment]
Looks like it passes except one build failed on installing matplotlib, (?) not sure if there's something I can do about that.

**Comment by twiecki** [comment]
Yeah, not sure why that sometimes failed. You can just restart the build.

On Tue, Nov 12, 2013 at 3:17 PM, chadheyne notifications@github.com wrote:

> Looks like it passes except one build failed on installing matplotlib, (?)
> not sure if there's something I can do about that.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/pymc-devs/pymc/pull/385#issuecomment-28329100
> .

## 

Thomas Wiecki
PhD candidate, Brown University
Quantitative Researcher, Quantopian Inc, Boston

**Comment by twiecki** [comment]
Also, can you enable both in travis, python 2.7 and 3.3?

**Comment by chadheyne** [comment]
I don't think I can tell Travis to restart since I'm not the owner. I did enable Python 2 and it tests okay. One of the Python 3 builds timed out during Numpy installation this time, though. :/

**Comment by fonnesbeck** [comment]
I just restarted the one that failed.

**Comment by chadheyne** [comment]
Looks like it passes tests on both Python 2 and Python 3. Anything else that you need me to do?

File: pymc/tests/test_step.py [hunk 1]
```diff
@@ -5,7 +5,7 @@
 
 
 def check_stat(name, trace, var, stat, value, bound):
-    s = stat(trace[var][4000:], axis=0)
+    s = stat(trace[var], axis=0)
```
Comment (twiecki) [inline pymc/tests/test_step.py]: Was this change required?

**Comment by twiecki** [comment]
Just went through this and it looks good. But see the comment I left. Otherwise good to merge I think.

File: pymc/tests/test_step.py (hunk 1 above)
Comment (chadheyne) [inline pymc/tests/test_step.py]: I changed it back. I don't remember why I changed it in the first place, to be honest. I just realized that I had closed the request earlier when commenting. Not what I meant to do, thanks for reopening.

**Comment by twiecki** [comment]
Awesome, thanks for helping out!
